### PR TITLE
Initial generics support

### PR DESCRIPTION
**Pull request by HaoK**
Hand merge with some changes from https://github.com/aspnet/Identity/pull/643

### CODE REVIEW

File: test/Microsoft.AspNetCore.Identity.EntityFrameworkCore.InMemory.Test/InMemoryStoreWithGenericsTest.cs [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+            IdentityResultAssert.IsSuccess(await manager.ReplaceClaimAsync(user, oldClaim, claim));
+            var newUserClaims = await manager.GetClaimsAsync(user);
+            Assert.Equal(1, newUserClaims.Count);
+            Claim newClaim = newUserClaims.FirstOrDefault();
+            Assert.Equal(claim.Type, newClaim.Type);
+            Assert.Equal(claim.Value, newClaim.Value);
+            Assert.Equal(claim.Issuer, newClaim.Issuer);
+        }
+
+        [Fact]
```
Comment (HaoK) [inline test/Microsoft.AspNetCore.Identity.EntityFrameworkCore.InMemory.Test/InMemoryStoreWithGenericsTest.cs]: @divega This test depends on the ability to affect querys using the "logincontext" to affect the filter

File: test/Microsoft.AspNetCore.Identity.EntityFrameworkCore.InMemory.Test/InMemoryStoreWithGenericsTest.cs [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+
+        protected override Expression<Func<IdentityUserWithGenerics, bool>> UserNameEqualsPredicate(string userName) => u => u.UserName == userName;
+
+        protected override Expression<Func<MyIdentityRole, bool>> RoleNameEqualsPredicate(string roleName) => r => r.Name == roleName;
+
+        protected override Expression<Func<IdentityUserWithGenerics, bool>> UserNameStartsWithPredicate(string userName) => u => u.UserName.StartsWith(userName);
+
+        protected override Expression<Func<MyIdentityRole, bool>> RoleNameStartsWithPredicate(string roleName) => r => r.Name.StartsWith(roleName);
+
+        [Fact]
```
Comment (HaoK) [inline test/Microsoft.AspNetCore.Identity.EntityFrameworkCore.InMemory.Test/InMemoryStoreWithGenericsTest.cs]: This test also requires modifying a query

File: test/Microsoft.AspNetCore.Identity.EntityFrameworkCore.InMemory.Test/InMemoryStoreWithGenericsTest.cs (hunk 1 above)
Comment (HaoK) [inline test/Microsoft.AspNetCore.Identity.EntityFrameworkCore.InMemory.Test/InMemoryStoreWithGenericsTest.cs]: An alternative way for this scenario to work without exposing all of the querys, would be to extend the user manager and overriding the behavior of the methods to do the right thing, either via new store api or directly applying a query.

File: src/Microsoft.AspNetCore.Identity.EntityFrameworkCore/UserStore.cs [hunk 3]
```diff
@@ -99,6 +157,65 @@ public UserStore(TContext context, IdentityErrorDescriber describer = null)
         public IdentityErrorDescriber ErrorDescriber { get; set; }
 
         /// <summary>
+        /// Returns the DbSet of roles.
+        /// </summary>
+        protected DbSet<TRole> Roles { get { return Context.Set<TRole>(); } }
```
Comment (HaoK) [inline src/Microsoft.AspNetCore.Identity.EntityFrameworkCore/UserStore.cs]: back to private on these

File: src/Microsoft.AspNetCore.Identity.EntityFrameworkCore/UserStore.cs [hunk 4]
```diff
@@ -109,7 +226,7 @@ public UserStore(TContext context, IdentityErrorDescriber describer = null)
         /// <summary>Saves the current store.</summary>
         /// <param name="cancellationToken">The <see cref="CancellationToken"/> used to propagate notifications that the operation should be canceled.</param>
         /// <returns>The <see cref="Task"/> that represents the asynchronous operation.</returns>
-        private Task SaveChanges(CancellationToken cancellationToken)
```
Comment (HaoK) [inline src/Microsoft.AspNetCore.Identity.EntityFrameworkCore/UserStore.cs]: back to private

File: test/Microsoft.AspNetCore.Identity.EntityFrameworkCore.InMemory.Test/InMemoryContext.cs [hunk 5]
```diff
@@ -35,4 +34,23 @@ protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
...
+        where TUserClaim : IdentityUserClaim<TKey>
+        where TUserRole : IdentityUserRole<TKey>
+        where TUserLogin : IdentityUserLogin<TKey>
+        where TRoleClaim : IdentityRoleClaim<TKey>
+        where TUserToken : IdentityUserToken<TKey>
+    {
+        public InMemoryContext(DbContextOptions options) : base(options)
+        { }
+
+        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
```
Comment (HaoK) [inline test/Microsoft.AspNetCore.Identity.EntityFrameworkCore.InMemory.Test/InMemoryContext.cs]: Probably unneeded

File: test/Microsoft.AspNetCore.Identity.EntityFrameworkCore.InMemory.Test/InMemoryStoreWithGenericsTest.cs [hunk 6]
```diff
@@ -0,0 +1,353 @@
...
+            var manager = CreateManager();
+            var user = CreateTestUser();
+            IdentityResultAssert.IsSuccess(await manager.CreateAsync(user));
+            Claim[] claims = { new Claim("c1", "v1", null, "i1"), new Claim("c2", "v2", null, "i2"), new Claim("c2", "v3", null, "i3") };
+            foreach (Claim c in claims)
+            {
+                IdentityResultAssert.IsSuccess(await manager.AddClaimAsync(user, c));
+            }
+
+            var userId = await manager.GetUserIdAsync(user);
```
Comment (HaoK) [inline test/Microsoft.AspNetCore.Identity.EntityFrameworkCore.InMemory.Test/InMemoryStoreWithGenericsTest.cs]: Scope these tests down again

**Comment by HaoK** [comment]
@divega This PR has been updated as we discussed, I removed the test that was trying to customize and changing the uniqueness of our login providers, the other 2 tests pass.  I also copied the test over to the SqlServer, and those pass as well, so I think this should be good to go.

**Comment by slaneyrw** [comment]
Great to see this is now getting traction... thanks for listening

File: src/Microsoft.AspNetCore.Identity.EntityFrameworkCore/IdentityUserClaim.cs [hunk 7]
```diff
@@ -30,5 +31,24 @@ public class IdentityUserClaim<TKey> where TKey : IEquatable<TKey>
...
+        public virtual Claim ToClaim()
+        {
+            return new Claim(ClaimType, ClaimValue);
+        }
+
+        /// <summary>
+        /// Reads the type and value from the Claim.
+        /// </summary>
+        /// <param name="claim"></param>
+        public virtual void FromClaim(Claim claim)
```
Comment (HaoK) [inline src/Microsoft.AspNetCore.Identity.EntityFrameworkCore/IdentityUserClaim.cs]: Initialize

**Comment by HaoK** [comment]
104f2161915b1a4e7d0013f3de3b8d9a212a5d27

**Pull request closed without merging**